structuredheadings/plugin.js

```diff
@@ -1,67 +1,139 @@
 (function () {
   CKEDITOR.plugins.add("structuredheadings", {
-    icons: "structuredheadings",
+    icons: "autonumberheading, matchheading",
     init: function (editor) {
       // list of elements allowed to be numbered
       var allowedElements = ["h1", "h2", "h3", "h4", "h5", "h6"];
 
       editor.addContentsCss(this.path + "styles/numbering.css");
 
       // Helper function to set button state
-      var setCommandState = function (state) {
+      var setCommandState = function (command, state) {
         switch (state) {
         case "on":
-          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_ON);
+          editor.getCommand(command).setState(CKEDITOR.TRISTATE_ON);
           break;
         case "off":
-          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_OFF);
+          editor.getCommand(command).setState(CKEDITOR.TRISTATE_OFF);
           break;
         case "disabled":
-          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_DISABLED);
+          editor.getCommand(command).setState(CKEDITOR.TRISTATE_DISABLED);
           break;
         }
 
       };
 
+      var isNumbered = function (element) {
+        if (element.hasClass("autonumber")) {
+          return true;
+        } else {
+          return false;
+        }
+      };
+
       // The command to add an autonumbering class to selection
       editor.addCommand("autoNumberHeading", {
         allowedContent: "h1(*); h2(*); h3(*); h4(*); h5(*); h6(*)",
-        requiredContent: allowedElements.join(";"),
         startDisabled: true,
         exec: function () {
           var element = editor.getSelection().getStartElement();
 
-          if (!element.hasClass("autonumber")) {
+          if (!isNumbered(element)) {
             element.addClass("autonumber");
-            setCommandState("on");
+            setCommandState("autoNumberHeading", "on");
           } else {
             element.removeClass("autonumber");
-            setCommandState("off");
+            setCommandState("autoNumberHeading", "off");
           }
         }
       });
 
+      editor.addCommand("matchHeading", {
+        startDisabled: true,
+        exec: function () {
+          var element = editor.getSelection().getStartElement();
```

Comment:
I'm not totally sure now either.  A range doesn't necessarily "break" things but it can introduce some ambiguity if someone selects multiple tags ex:

```
<p>Hea[ding 1</p>
<p>Heading 2]</p>
```

Another thing I just thought of - getStartElement may misbehave in cases like
```
<p>Something<strong>^asdf</strong></p>
```

I'm not totally sure the below (getPrevious) works as expected in this case